src/flist/mount.rs

```diff
@@ -0,0 +1,487 @@
+            .filter(|mnt_info| mnt_info.target.starts_with(&self.root))
+            .filter(|mnt_info| {
+                mnt_info.target.parent() == Some(&self.ro)
+                    && mnt_info.filesystem == FsType::G8UFS.as_ref()
+            });
+
+        for mount in ros {
+            let pid: i64 = mount.source.parse()?;
+            ro_targets.insert(pid, mount);
+        }
+
+        let all_under_mountpoints = all
+            .iter()
+            .filter(|mount| mount.target.parent() == Some(&self.mountpoint));
+
+        for mount in all_under_mountpoints {
+            let pid: i64;
+            if mount.filesystem == FsType::G8UFS.as_ref() {
+                pid = mount.source.parse()?
+            } else if mount.filesystem == FsType::Overlay.as_ref() {
+                // let lower_dir_path = mount.as_overlay()?.lower_dir;
+                let lower_dir = match mount.option("lowerdir") {
+                    Some(Some(lower_dir)) => lower_dir,
+                    _ => bail!("bad overlay options: lowerdir not found"),
+                };
+                let mut all_matching_overlay = all
+                    .iter()
+                    .filter(|mnt| PathBuf::from(lower_dir) == mnt.target);
+                pid = match all_matching_overlay.next() {
+                    Some(mount) => mount.source.parse()?,
+                    None => continue,
+                };
+            } else {
+                continue;
+            }
+            ro_targets.remove(&pid);
+        }
+        for (_, mount) in ro_targets.iter() {
+            log::debug!("cleaning up mount {}", &mount.target.display());
+            if let Err(err) = self.syscalls.umount(&mount.target, None) {
+                log::debug!(
+                    "failed to unmount {} Error: {}",
+                    mount.target.display(),
+                    err
+                );
+                continue;
+            }
+            if let Err(err) = fs::remove_dir_all(&mount.target).await {
+                log::debug!("failed to remove dir {:#?}  Error: {}", mount.target, err);
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn get_volume_path<T: AsRef<str>>(&self, name: T, size: u64) -> Result<PathBuf> {
+        // no persisted volume provided, hence
+        // we need to create one, or find one that is already exists
+        match self.storage.volume_lookup(&name) {
+            Ok(volume) => Ok(volume.path),
+            Err(_) => {
+                // Volume doesn't exist create a new one
+                if size == 0 {
+                    bail!("invalid mount option, missing disk type");
+                }
+                match self.storage.volume_create(&name, size) {
+                    Ok(volume) => Ok(volume.path),
+                    Err(e) => {
+                        self.storage.volume_delete(&name)?;
+                        bail!(e)
+                    }
+                }
+            }
+        }
+    }
```

Comment:
I think this is also not clean enough:
- First of all the check for size (if size == 0) can be at the header of the function. not deep inside the match .
- You can do an early return directly like 
```rust
if let Ok(volume) == self.storage.volume_lookup(&name) {
   return Ok(volume);
}
// then you try the create 
```